### PR TITLE
3151 scoped threads 翻譯

### PR DESCRIPTION
**Pull request by FizzyElt**
目前有幾個問題想請教
- 關於 commit 連結我發現有很多個不知道該挑哪一個（原本的編號好像是 0000 ?）
- join threads 翻成 "加入執行緒" 我覺得有點奇怪，不知是否有更好的解釋

### CODE REVIEW

**Comment by weihanglo** [comment]
> * 關於 commit 連結我發現有很多個不知道該挑哪一個（原本的編號好像是 0000 ?）

Perhaps this one? https://github.com/rust-lang/rfcs/commit/6de9eb35193852dd910c38f5c2c41e6f3e2a8ed4

> * join threads 翻成 "加入執行緒" 我覺得有點奇怪，不知是否有更好的解釋

According to [《UNIX 作業系統 fork/exec 系統呼叫的前世今生》](https://hackmd.io/@sysprog/unix-fork-exec), it says _"join 表示多個並行處理的行程由於某種原因不得不進行同步的時間點，也就是多個並行行程**會合**之處"_. So, it might be "會合". I have no preference, either.

**Review by weihanglo — COMMENTED** [review]
There are some pre-existing translations for each title of chapter, such as **Guide-level explanation -> 教學式解說**. I lean towards unifying them a bit. We surely can translate them better afterwards.

For the references:

- https://rust-lang.tw/rfcs-tw/2342-const-control-flow.html
- https://rust-lang.tw/rfcs-tw/3128-io-safety.html

File: text/3151-scoped-threads.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+- RFC PR: [rust-lang/rfcs#3151](https://github.com/rust-lang/rfcs/pull/3151)
+- Rust Issue: [rust-lang/rust#93203](https://github.com/rust-lang/rust/issues/93203)
+- Translators: [[@FizzyElt](https://github.com/FizzyElt)]
+- Commit: [The commit link this page based on](https://github.com/rust-lang/rfcs/blob/6de9eb35193852dd910c38f5c2c41e6f3e2a8ed4/text/3151-scoped-threads.md)
+- Updated: 2022-07-05
+
+# 概要
+[概要]: #summary
+
+在標準函式庫中新增作用域執行緒，其允許借用父執行續的變數來產生執行緒。
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]: The translation is not accurate, I propose the following:

```suggestion
在標準函式庫中新增作用域執行緒，其允許產生可借用父執行緒之變數的執行緒。
```

File: text/3151-scoped-threads.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    s.spawn(|_| println!("borrowed from thread #2: {}", var));
+});
+```
+
+# 動機
+[動機]: #motivation
+
+在 Rust 1.0 發布之前，我們有與作用域執行續相同作用的 [`thread::scoped()`](https://docs.rs/thread-scoped/1.0.2/thread_scoped/)，但後來發現一個健全性的問題，可能導致 use-after-frees，所以它已被移除。這一歷史事件稱為[洩密事件](http://cglab.ca/~abeinges/blah/everyone-poops/)。
+
+幸運的是，舊的作用域執行緒可以透過依靠 closure 而不是守護來確保生成的執行緒被自動加入。但我們對在 Rust 1.0 中加入的作用域執行緒並不放心，所以我們決定將其放在外部 crates 之中，並有可能在未來的某個時刻回到標準函式庫中。四年過去了，那個未來就是現在。
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]: I would expect 守護 and 加入 can have an auxiliary note for their original words, such like "守護（guard）".

In addition, the first sentense here is not accurate. I read it as "幸運的是，舊的作用域執行緒可被修復，透過閉包（closure）而非守護（guard)，來確保生成的執行緒會自動會合（join）。

File: text/3151-scoped-threads.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+let greeting = String::from("Hello world!");
+
+let handle = thread::spawn(move || {
+    println!("thread #1 says: {}", greeting);
+});
+
+handle.join().unwrap();
+```
+
+現在讓我們嘗試生成兩個使用相同 greeting 的線程。不幸的是，我們必須克隆它，因為 [`thread::spawn()`](https://doc.rust-lang.org/std/thread/fn.spawn.html) 有 `F: 'static` 要求，這意味著執行緒不能借用局部變數：
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]:
```suggestion
現在讓我們嘗試生成兩個使用相同 greeting 的執行緒。不幸的是，我們必須克隆它，因為 [`thread::spawn()`](https://doc.rust-lang.org/std/thread/fn.spawn.html) 有 `F: 'static` 要求，這意味著執行緒不能借用局部變數：
```

File: text/3151-scoped-threads.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    });
+
+    handle1.join().unwrap();
+    handle2.join().unwrap();
+});
+```
+
+這意味著可以毫無顧忌地借用作用域之外的變數！
+
+現在我們不必再手動加入執行緒，因為所有未加入的執行緒將在作用域結束時自動加入：
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]: So you choose to translate "join" into 加入, right? Now I feel like "會合" is a better term as it implies all threads eventually converge.

File: text/3151-scoped-threads.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+    s.spawn(|_| {
+        println!("thread #2 says: {}", greeting);
+    });
+});
+```
+
+當以這種方式利用自動加入時，請注意，如果任何自動加入的執行緒出現恐慌，`thread::scope()` 將出現恐慌。
+
+你可能已經註意到作用域執行緒現在只接受一個參數，它只是對 `s` 的另一個引用。由於 `s` 存在於作用域內，我們不能直接借用它。使用傳遞的參數來生成巢狀執行緒：
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]:
```suggestion
你可能已經注意到作用域執行緒現在只接受一個參數，它只是對 `s` 的另一個引用。由於 `s` 存在於作用域內，我們不能直接借用它。使用傳遞的參數來生成巢狀執行緒：
```

File: text/3151-scoped-threads.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+  * 這是一個非常常見和實用的工具，非常適合學習、測試和探索性程式設計。每個學習 Rust 的人都會在某個時候遇到借用和執行緒的互動。有一個非常重要的教訓是執行緒實際上**可以**借用局部變數，但標準函式庫並沒有反映這一點。
+
+  * 有些人可能會爭辯說我們應該完全不鼓勵使用執行緒，而是將人們指向像 Rayon 和 Tokio 這樣的執行器。但是，`thread::spawn()` 需要 F: `F: 'static` 並且無法繞過它，這感覺就像是標準函式庫中缺少的部分。
+
+  * 實現作用域執行緒非常難處理，因此最好有標準函式庫提供一個可靠的解決方案。
+
+  * 官方文檔和書籍中有許多範例可以透過作用域執行緒進行簡化。
+
+  * 作用域執行緒通常是比 `thread::spawn()` 更好的默認值，因為它們確保生成的執行緒被連接並且不會意外 “洩漏”。這有時是單元測試中的一個問題，如果單元測試產生執行緒並忘記加入它們，“迷途” 的執行緒可能會積累。
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]: * Prefer 「」 over informal punctuations in Traditional Chinese.
* "連接" here is also "join". I expect that should aligns to translations in other places.

```suggestion
  * 作用域執行緒通常是比 `thread::spawn()` 更好的預設值，因為它們確保生成的執行緒被連接並且不會意外「洩漏」。這有時是單元測試中的一個問題，如果單元測試產生執行緒並忘記加入它們，「迷途」的執行緒可能會積累。
```

File: text/3151-scoped-threads.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+3. 我們刪除了 `Scope::defer()`，因為它不是真的有用，有錯誤，並且有不明顯的行為。
+
+4. `ScopedJoinHandle` 在 `'scope` 上進行了參數化，以防止它逃離作用域。
+
+Rayon 也有作用域，但它們在不同的抽象級別上工作——Rayon 產生任務而不是執行緒。它的 API 與本 RFC 中提出的 API 相同。
+
+# 未解決的問題
+[未解決的問題]: #unresolved-questions
+
+這個概念可以延伸到異步嗎？會有任何行為或 API 差異嗎？
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]:
```suggestion
這個概念可以延伸到非同步嗎？會有任何行為或 API 差異嗎？
```

**Review by weihanglo — COMMENTED** [review]
Thank you for your translation! Some places need tweaks and feel free to discuss.

**Comment by FizzyElt** [comment]
已修正

File: text/3151-scoped-threads.md (hunk 1 above)
Comment (weihanglo) [reply text/3151-scoped-threads.md]: Sorry but your new change is still not quite right. We are not borrowing variables to spawn a new thread. We are spawning threads that allow to borrow variables.

File: text/3151-scoped-threads.md (hunk 4 above)
Comment (weihanglo) [reply text/3151-scoped-threads.md]: Can we maintain the cohesion of translation terms? At this moment, both of "加入" and "會合" exist in the change. I would like to see only one got picked :)

File: text/3151-scoped-threads.md [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+  * 有些人可能會爭辯說我們應該完全不鼓勵使用執行緒，而是將人們指向像 Rayon 和 Tokio 這樣的執行器。但是，`thread::spawn()` 需要 F: `F: 'static` 並且無法繞過它，這感覺就像是標準函式庫中缺少的部分。
+
+  * 實現作用域執行緒非常難處理，因此最好有標準函式庫提供一個可靠的解決方案。
+
+  * 官方文檔和書籍中有許多範例可以透過作用域執行緒進行簡化。
+
+  * 作用域執行緒通常是比 `thread::spawn()` 更好的預設值，因為它們確保生成的執行緒被連接並且不會意外 「洩漏」。這有時是單元測試中的一個問題，如果單元測試產生執行緒並忘記加入它們，「迷途」 的執行緒可能會積累。
+
+  * 使用者一直在 IRC 和論壇上詢問作用域執行緒。將它們作為 `std::thread` 中的“祝福”模式對每個人都有好處。
```
Comment (weihanglo) [inline text/3151-scoped-threads.md]: I missed this quotation mark. Should change to 「」 as well. Thanks!

**Review by weihanglo — COMMENTED** [review]
Sorry being nitpicking. And thank you for your patient to translate this RFC!

File: text/3151-scoped-threads.md (hunk 4 above)
Comment (FizzyElt) [reply text/3151-scoped-threads.md]: Sorry, I didn't notice 😢

File: text/3151-scoped-threads.md (hunk 8 above)
Comment (FizzyElt) [reply text/3151-scoped-threads.md]: It's okay, this is my negligence

**Comment by FizzyElt** [comment]
-  加入 -> 會合
- "祝福" -> 「祝福」

**Comment by FizzyElt** [comment]
fixed
- 加入 -> 會合
- "祝福" -> 「祝福」
- 其允許借用父執行續的變數來產生執行緒 -> 其允許產生可借用父執行緒之變數的執行緒

**Review by weihanglo — APPROVED** [review]
Looks pretty well. Thank you so much for the translation!

**Pull request merged**